Improvements

xAstroBoy opened the pull request:
Add /Suit command
Add /Enemies Command
Add /Spawn command
Add /quota command 

Fix Godmode to not break player interactions
add DelayedFunction extension
Fix kill Self not working by delaying the toggle on by a few seconds late giving time to execute.

winstxnhdw commented:
The branch is already a mess. If you look at the file changes, you are beginning to revert our current changes.

Create a new PR with one feature each, or you can wait for me to commit them myself.

xAstroBoy commented:
> The branch is already a mess. If you look at the file changes, you are beginning to revert our current changes.
> 
> Create a new PR with one feature each, or you can wait for me to commit them myself.

i am working on reverting the edits that are outdated.

winstxnhdw commented on lc-hax/Scripts/Patches/NoFallDamagePatch.cs:
Revert this

winstxnhdw replied:
This isn’t for God mode. It’s for FollowMod

xAstroBoy replied:
yeah and is merged under godmode.cs , i mentioned you where i copied the code? but ill reverse it as your request.

xAstroBoy commented on lc-hax/Scripts/Patches/GodModePatch.cs:
```diff
@@ -1,10 +1,123 @@
+    static bool PrefixDamagePlayer(PlayerControllerB __instance, CauseOfDeath causeOfDeath) {
+        if (__instance.isSelf()) {
+            if (Setting.EnableGodMode) return false;
+            if (Setting.DisableFallDamage && causeOfDeath == CauseOfDeath.Gravity) return false;
```
@winstxnhdw No Fall Damage is merged here.

xAstroBoy commented on lc-hax/Scripts/Patches/GodModePatch.cs:
```diff
@@ -1,10 +1,123 @@
+    static bool PrefixKillPlayer(PlayerControllerB __instance, CauseOfDeath causeOfDeath) {
+        if (__instance.isSelf()) {
+            if (Setting.EnableGodMode) return false;
+            if (Setting.DisableFallDamage && causeOfDeath == CauseOfDeath.Gravity) return false;
```
No Fall Damage is merged here as well.

winstxnhdw commented:
Move everything to do with EnemyController to another PR.

xAstroBoy commented:
> Move everything to do with EnemyController to another PR.

there's not a lot of different changes, just analyzed each AI and converted their behaviour states to enumerators which makes it easy to understand and adjusted the way to make RigidBodyKeyboard set the enemy walking/running animation based off the given imputs.
i forgot to merge this before you fixed and cleaned the code properly.

winstxnhdw commented:
> there's not a lot of different changes

Exactly, that’s why move it to another PR. I am not reviewing another PR with dozens of things going on again.

xAstroBoy commented:
@winstxnhdw im sure now the files edited aren't even editing the original codes anymore, i triple checked everything.

xAstroBoy commented:
@winstxnhdw i removed all the obsolete edits , just the needed changes.

winstxnhdw commented on lc-hax/Scripts/Helpers/Unity.cs:
Why can't you use `Helper.CreateComponent<WaitForBehaviour>()` instead?

winstxnhdw replied:
For duration, you can do it like this.

```cs
        Helper.CreateComponent<WaitForBehaviour>("Stop Pulling Horn")
              .SetPredicate(time => time >= hornDuration)
              .Init(() => shipAlarmCord?.StopPullingCordServerRpc(-1));
```

xAstroBoy replied:
> For duration, you can do it like this.
> 
> ```cs
>         Helper.CreateComponent<WaitForBehaviour>("Stop Pulling Horn")
>               .SetPredicate(time => time >= hornDuration)
>               .Init(() => shipAlarmCord?.StopPullingCordServerRpc(-1));
> ```

the issue is that it spawns a lot of them, for example /mask it spawns like when you want one it spawns 20.

winstxnhdw replied:
What?

xAstroBoy replied:
> Why can't you use `Helper.CreateComponent<WaitForBehaviour>()` instead?

because there's no need as is just a simply delayed coroutine that invokes a action.

winstxnhdw replied:
You're right. That's not intended though. I'll take a look.

xAstroBoy replied:
Speaking about that, since you use one in /spawn, make sure is just one as a person can type /revive as many times if they die multiple times.

winstxnhdw replied:
Btw, the one in `/mask` is `TransientBehaviour`, not `WaitForBehaviour` which is basically the same as your implementation.

winstxnhdw replied:
> Speaking about that, since you use one in /spawn, make sure is just one as a person can type /revive as many times if they die multiple times.

Sure, `CreateComponent` has a built-in pool, so I can automatically limit it to only one game object at a time.

winstxnhdw replied:
Okay, I fixed the mask issue. Please use `WaitForBehaviour` instead. It's heavily optimised, works outside of lobbies, has a built-in queue, and it does the same thing.

winstxnhdw commented:
Also, is your ReSharper not detecting my editorconfig? I see a lot of format style commits and that happens if you keep violating the editorconfig settings. Normally, it should help you autoformat the code.

xAstroBoy commented:
> Also, is your ReSharper not detecting my editorconfig? I see a lot of format style commits and that happens if you keep violating the editorconfig settings. Normally, it should help you autoformat the code.

apparently so...

xAstroBoy commented:
@winstxnhdw i turned that predicate into a extension

xAstroBoy commented:
Looks solid to me.

winstxnhdw reviewed (commented):
Is this necessary? `KillPlayer` should be synchronous and atomic, so it'll run immediately.

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/Controllers/NutcrackerController.cs:
```diff
@@ -15,9 +21,19 @@ internal class NutcrackerController : IEnemyController<NutcrackerEnemyAI> {
 
-    public void UseSecondarySkill(NutcrackerEnemyAI enemyInstance) => enemyInstance.SetBehaviourState(NutCrackerState.SENTRY);
+    public void UseSecondarySkill(NutcrackerEnemyAI enemyInstance) {
+        if (!this.isSecondarySkillActive)
+            enemyInstance.SetBehaviourState(NutCrackerState.SENTRY);
```
We use braces for multi-line if-statements.

winstxnhdw commented on lc-hax/Scripts/Loader.cs:
```diff
@@ -76,6 +77,11 @@ internal class Loader : MonoBehaviour {
         Loader.AddHaxModules<InstantInteractMod>();
+
+        // Dont Disable this line, it is used to fill the Helper in case is injected during a full loaded level.
+        Loader.AddHaxModules<RefreshMod>();
```
Let's move this to `AddHaxGameObject` instead, then move `RefreshMod` to the `Core` directory and rename it to `Refresh`.

winstxnhdw commented on lc-hax/Scripts/Helpers/Enemies.cs:
```diff
@@ -29,4 +77,19 @@ internal static partial class Helper {
 
         return null;
     }
+    internal static void MouthDogChasePlayer(this MouthDogAI instance, PlayerControllerB player) {
+        if (instance == null) return;
+        if (player == null) return;
+        if (instance.currentBehaviourStateIndex is 0 or 1) {
+            _ = instance.Reflect().InvokeInternalMethod("ChaseLocalPlayer");
+        }
+        else {
+            if (instance.currentBehaviourStateIndex != 2 || instance.Reflect().GetInternalField<bool>("inLunge"))
+                return;
+            instance.transform.LookAt(player.transform.position);
+            instance.transform.localEulerAngles = new Vector3(0.0f, player.transform.eulerAngles.y, 0.0f);
+            _ = instance.Reflect().SetInternalField("inLunge", true);
+            _ = instance.Reflect().InvokeInternalMethod("EnterLunge");
+        }
+    }
```
Make this inline. No point abstracting something we will never use. Even if we do want to use this, it doesn't follow SRP anyways.

xAstroBoy replied:
is being used in god mode , to make the dog have a controlled interaction with the local player.

winstxnhdw replied:
ya so just write it inline there. don’t add this to helper.

winstxnhdw replied:
you are polluting the helper namespace

xAstroBoy replied:
All done, moved it to the patch directly.

winstxnhdw commented on lc-hax/Scripts/Helpers/Players.cs:
```diff
@@ -37,4 +37,6 @@ internal static partial class Helper {
 
     internal static PlayerControllerB? GetActivePlayer(int playerClientId) => Helper.GetActivePlayer(playerClientId.ToString());
+
+    internal static bool isSelf(this PlayerControllerB instance) => Helper.LocalPlayer != null && instance.actualClientId == Helper.LocalPlayer.actualClientId;
```
Methods should be `PascalCase`. This piece of code is used often, so it's worth to optimise it.

```cs
[MethodImpl(MethodImplOptions.AggressiveInlining)]
internal static bool isSelf(this PlayerControllerB instance) => Helper.LocalPlayer is PlayerControllerB localPlayer && instance.actualClientId == localPlayer.actualClientId;
```

winstxnhdw commented on lc-hax/Scripts/Modules/RefreshMod.cs:
```diff
@@ -0,0 +1,35 @@
+
+    private void Awake() {
+        if (Helper.LocalPlayer is null) return;
+        if (Helper.StartOfRound is null) return;
+        if (Helper.StartOfRound.shipHasLanded) {
```
Use null-condtional here.

winstxnhdw commented on lc-hax/Scripts/Modules/RefreshMod.cs:
```diff
@@ -0,0 +1,35 @@
+        GrabbableObject[] Scraps = Helper.FindObjects<GrabbableObject>();
+        foreach (GrabbableObject Scrap in Scraps) {
+            // check if they are actively online.
+            if (Scrap.IsSpawned) {
+                _ = Helper.Grabbables.Add(Scrap);
+            }
```
if-guard instead

winstxnhdw commented on lc-hax/Scripts/Modules/RefreshMod.cs:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private void FillScrapList() {
+        GrabbableObject[] Scraps = Helper.FindObjects<GrabbableObject>();
+        foreach (GrabbableObject Scrap in Scraps) {
```
Just use `.ForEach` here

winstxnhdw commented on lc-hax/Scripts/Patches/GodModePatch.cs:
```diff
@@ -1,10 +1,123 @@
+    [HarmonyPrefix]
+    [HarmonyPatch(typeof(PlayerControllerB), nameof(PlayerControllerB.AllowPlayerDeath))]
+    static bool PrefixDamagePlayer(PlayerControllerB __instance, ref bool __result) {
+        if (__instance.isSelf())
```
if-guard

xAstroBoy replied:
is THE SAME!

winstxnhdw replied:
just because the logic is the same, doesn’t mean it’s good. nesting forces the user to read the conditional when otherwise, it doesn’t have to be read.

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/QuotaCommand.cs:
```diff
@@ -0,0 +1,37 @@
+
+    void SetQuota(int profit, int fulfilled = 0) {
+        if (TimeOfDay.Instance == null) return;
+        TimeOfDay.Instance.profitQuota = profit;
+        TimeOfDay.Instance.quotaFulfilled = fulfilled;
+        TimeOfDay.Instance.UpdateProfitQuotaCurrentTime();
```
Move to `Managers` then we can use it in `ClearVision` too.

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/QuotaCommand.cs:
```diff
@@ -0,0 +1,37 @@
+
+    public void Execute(StringArray args) {
+        if (args.Length < 2) {
+            Chat.Print("Usage: /quota <amount> <?fulfilled>");
```
```suggestion
            Chat.Print("Usage: /quota <amount> <fulfilled?>");
```

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/QuotaCommand.cs:
```diff
@@ -0,0 +1,37 @@
+            return;
+        }
+
+        if (!int.TryParse(args[0], out int amount)) {
```
Can amount be < 0? Else use unsigned integers.

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/SpawnCommand.cs:
```diff
@@ -0,0 +1,67 @@
+    private void SpawnEnemyOnPlayer(PlayerControllerB player, GameObject prefab, ulong amount = 1) {
+        if (Helper.RoundManager == null) return;
+        if (player == null) return;
+        for (ulong i = 0; i < amount; i++) {
+            GameObject enemy = UnityEngine.Object.Instantiate(prefab, player.transform.position, Quaternion.Euler(Vector3.zero));
+            if (enemy != null) {
+                if (enemy.TryGetComponent(out NetworkObject Network)) {
+                    Network.Spawn(true);
+                    if (enemy.TryGetComponent(out EnemyAI AI)) {
+                        _ = Helper.Enemies.Add(AI);
+                    }
+                }
```
if-guard

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/SpawnCommand.cs:
```diff
@@ -0,0 +1,67 @@
+
+        string name = args[1];
+        string EnemyName = "";
+        GameObject? prefab = null;
+        foreach (KeyValuePair<string, GameObject> enemy in Helper.SpawnableEnemies) {
+            if (enemy.Key.Contains(name, StringComparison.OrdinalIgnoreCase)) {
+                EnemyName = enemy.Key;
+                prefab = enemy.Value;
+                break;
+            }
+        }
+        if (prefab == null) {
+            return;
+        }
```
refactor to use only Linq then you don't have to create so many separate variables.

xAstroBoy replied:
Both variables are needed because 
1) use the prefab
2) print the name of the enemy being spawned

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/LevelEnemiesCommand.cs:
```diff
@@ -0,0 +1,19 @@
+using System;
+using UnityEngine;
+
+[DebugCommand("/enemies")]
```
This is a debug command. Move it to `Debug`.

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/LevelEnemiesCommand.cs:
```diff
@@ -0,0 +1,19 @@
+        if (Helper.RoundManager == null) return;
+        if (Helper.RoundManager.currentLevel == null) return;
+        string Msg = "";
+        foreach (System.Collections.Generic.KeyValuePair<string, GameObject> enemy in Helper.SpawnableEnemies) {
+            Msg += $"{enemy.Key}, ";
+        }
```
Use `string.Join` instead.

xAstroBoy replied:
is the same!

winstxnhdw replied:
no it isn’t. string.Join only does the string concatenation once. your method does the concatenation N times.

xAstroBoy replied:
Yeah i fixed that.

winstxnhdw reviewed (changes requested):
No, not solid.

winstxnhdw commented:
Move the enemy controller stuff to another PR if you want that merged soon.

winstxnhdw commented on lc-hax/Scripts/Commands/PlayerDeathNotifyCommand.cs:
I rather this be a mod that displays the dead players on the top-left honestly.

xAstroBoy replied:
this spawns a hud notification, but it can be enhanced later with the minimal gui.

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/QuotaCommand.cs:
Use `ushort` instead and you won’t have to cast.

xAstroBoy replied:
cast is still needed, only int

winstxnhdw commented:
can you please stop adding new things to the PR. just create a new one. it’s frustrating to review code that is constantly changing. it’ll also force you to think how you can separate code better to avoid merge conflicts, which will help you with structuring your code better.

xAstroBoy commented:
> can you please stop adding new things to the PR. just create a new one. it’s frustrating to review code that is constantly changing. it’ll also force you to think how you can separate code better to avoid merge conflicts, which will help you with structuring your code better.

that was the last one, im enchancing /spawn to cover all enemies of all levels, not just the ones based off that level, since controlcompany does a similiar thing with a override mode, i rewrote the dictionary parsing to look better and cleaner and it will fill it once as it will find all the enemies and prefabs in all levels.

xAstroBoy commented:
@winstxnhdw changes are done and finished.

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/QuotaCommand.cs:
```diff
@@ -0,0 +1,37 @@
+[PrivilegedCommand("/quota")]
+public class QuotaCommand : ICommand {
+
+    void SetQuota(ulong profit, ulong fulfilled = 0) {
+        if (Helper.TimeOfDay != null) {
+            Helper.TimeOfDay.profitQuota = (int)profit;
+            Helper.TimeOfDay.quotaFulfilled = (int)fulfilled;
+            Helper.TimeOfDay.UpdateProfitQuotaCurrentTime();
+        }
+    }
+
+    public void Execute(StringArray args) {
+        if (Helper.TimeOfDay == null) {
+            Chat.Print("TimeOfDay helper is not available.");
+            return;
+        }
+
+        if (args.Length < 1) {
+            Chat.Print("Usage: /quota <amount> <fulfilled?>");
+            return;
+        }
+
+        if (!ulong.TryParse(args[0], out ulong amount)) {
+            Chat.Print("Invalid amount!");
+            return;
+        }
+
+        ulong fulfilled = 0;
+        if (args.Length > 1 && !ulong.TryParse(args[1], out fulfilled)) {
+            Chat.Print("Invalid fulfilled amount, setting to 0.");
+        }
+
+        this.SetQuota(amount, fulfilled);
+    }
+}
```
Use `ushort` so you don't have to cast. Also, do `if (Helper.TimeOfDay is not TimeOfDay timeOfDay)` so you don't have to keep accessing `Helper.TimeOfDay` so many times.

xAstroBoy replied:
ushort still requires casting!

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/SpawnCommand.cs:
```diff
@@ -0,0 +1,66 @@
+                }
+            }
+            else {
+                UnityEngine.Object.Destroy(enemy);
```
```cs
using UnityObject = UnityEngine.Object

UnityObject.Destroy(enemy)
```

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/SpawnCommand.cs:
```diff
@@ -0,0 +1,66 @@
+        }
+
+        // Get the target player
+        PlayerControllerB? targetPlayer = Helper.GetActivePlayer(args[0]);
+        if (targetPlayer == null) {
```
```cs
if (Helper.GetActivePlayer(args[0]) is not PlayerControllerB targetPlayer) {
return;
}
```

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/SpawnCommand.cs:
```diff
@@ -0,0 +1,66 @@
+
+        string enemyNamePart = args[1];
+        KeyValuePair<string, GameObject> enemyEntry = Helper.AllSpawnableEnemies
+            .FirstOrDefault(e => e.Key.Contains(enemyNamePart, StringComparison.OrdinalIgnoreCase));
```
```suggestion
            .First(e => e.Key.Contains(enemyNamePart, StringComparison.OrdinalIgnoreCase));
```

winstxnhdw commented on lc-hax/Scripts/Core/Refresh.cs:
```diff
@@ -0,0 +1,22 @@
+public sealed class Refresh : MonoBehaviour {
+    private void Awake() {
+        if (Helper.LocalPlayer == null || Helper.StartOfRound == null) return;
+        if (Helper.StartOfRound.shipHasLanded) this.FillEnemyList();
```
```cs
        if (Helper.StartOfRound?.shipHasLanded is true) {
this.FillEnemyList();
}
```

winstxnhdw commented on lc-hax/Scripts/Helpers/Hud.cs:
```diff
@@ -0,0 +1,10 @@
+internal static partial class Helper {
+
+    internal static void DisplayNotification(string title, string Text, bool isWarning = false) {
+        if (Helper.HUDManager == null) return;
+        Helper.HUDManager.DisplayTip(title, Text, isWarning, false);
```
```suggestion
        Helper.HUDManager?.DisplayTip(title, Text, isWarning, false);
```

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/Controllers/EyelessDogController.cs:
```diff
@@ -8,6 +8,9 @@ enum MouthDog {
 }
 
 internal class EyelessDogController : IEnemyController<MouthDogAI> {
+
```
```suggestion
```

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/Controllers/NutcrackerController.cs:
```diff
@@ -6,7 +6,14 @@ enum NutCrackerState {
 
 internal class NutcrackerController : IEnemyController<NutcrackerEnemyAI> {
-    public void OnMovement(NutcrackerEnemyAI enemyInstance, bool isMoving, bool isSprinting) => enemyInstance.SetBehaviourState(NutCrackerState.WALKING);
+
```
```suggestion
```

winstxnhdw commented on lc-hax/Scripts/Patches/DeathNotifierPatch.cs:
```diff
@@ -0,0 +1,15 @@
+class DeathNotifierPatch {
+    static void Postfix(int playerId, int causeOfDeath) {
+        if (!Setting.EnablePlayerDeathNotifications) return;
+        if (Helper.GetPlayer((ulong)playerId) is PlayerControllerB player) {
```
```cs
        if (Helper.GetPlayer((ulong)playerId) is not PlayerControllerB player) return;
```

xAstroBoy commented:
@winstxnhdw all good now?

xAstroBoy commented:
@winstxnhdw Now?

winstxnhdw commented:
Nope. Still too much nonsense.

xAstroBoy commented:
> Nope. Still too much nonsense.

what is nonsense? I DONT GET IT WHAT IS THE ISSUE!?

winstxnhdw commented:
This is why I told you to separate the PRs so we can have an easier time merging. Now you are stuck here all because you are holding unnecessary state in your suit patch.

xAstroBoy commented:
> This is why I told you to separate the PRs so we can have an easier time merging. Now you are stuck here all because you are holding unnecessary state in your suit patch.

the unneccessary state is not unneccessary, it caches everyting , EDITS AND RESTORES THE STATE OF THE ORIGINAL SUIT and it works, __STATE CAN CONTAIN JUST ONE VARIABLE, NOT 3!

winstxnhdw commented:
Use a struct.

xAstroBoy commented:
> Use a struct.

Why overcomplicate it ?
it works! NO NEED FOR A STRUCT!

winstxnhdw commented:
Because `__state` only works between `Prefix` and `Postfix`.

xAstroBoy commented:
> Because `__state` only works between `Prefix` and `Postfix`. Your static fields are accessible throughout the class.

BECAUSE THE PATCH HAS TO SET ONE FIELD TO OVERRIDE THE PRIVATE CLASS FIELD BEFORE EXECUTING IT!

winstxnhdw commented:
How about I ask you a question? Why not, I just convert every single variable, method and class in my program to `static`. It's easier and apparently, 'less complicated' after all.

xAstroBoy commented:
> How about I ask you a question? Why not, I just convert every single variable, method and class in my program to `static`. It's easier and apparently, 'less complicated' after all.

These variables can and they are exposed, because there's a reason, not because there's not.

xAstroBoy commented:
> Sure, then why create classes or structs at all? Just store them as static fields, right? They have a reason.

Because they need to be accessible from the same patch externally!
Why you overcomplicate it? if you want to make it your way just do it, i dont get it why there's such a big deal just for 3 fields exposed to be edited externally!?

winstxnhdw commented:
Same thing can be said, why not just use a `struct`? You should read up on https://en.wikipedia.org/wiki/SOLID. Using a `struct` instead of a `static` field is trivial. They both take the same amount of time to implement so why don't you just choose the better one?

Am I asking you to spend hours to profile your code for micro performance gains? No. Just write better; least accessible, typesafe, immutable code.

winstxnhdw commented on lc-hax/Scripts/Commands/Privileged/SpawnCommand.cs:
```diff
@@ -0,0 +1,63 @@
+        if (player == null) return;
+
+        for (ulong i = 0; i < amount; i++) {
+            GameObject enemy = UnityEngine.Object.Instantiate(prefab, player.transform.position, Quaternion.Euler(Vector3.zero));
```
I don't understand how you see that you have left two different ways to write the same thing and think "let's merge this!"

```suggestion
            GameObject enemy = UnityObject.Instantiate(prefab, player.transform.position, Quaternion.Euler(Vector3.zero));
```

winstxnhdw commented on lc-hax/Scripts/Core/Refresh.cs:
```diff
@@ -0,0 +1,24 @@
+using System.Linq;
+using UnityEngine;
+
+public sealed class Refresh : MonoBehaviour {
```
```suggestion
internal sealed class Refresh : MonoBehaviour {
```

winstxnhdw commented on lc-hax/Scripts/Patches/GodModePatch.cs:
```diff
@@ -1,10 +1,125 @@
+    // needed because the dogs do break the player even on contact, which makes it impossible to grab items,
+    // we patch it to make a controlled interaction for this, avoiding the death under god mode.
+    [HarmonyPrefix]
+    [HarmonyPatch(typeof(MouthDogAI), nameof(MouthDogAI.OnCollideWithPlayer))]
+    static bool ControlledDogCollision(MouthDogAI __instance, ref Collider other) {
+        if (!Setting.EnableGodMode) return true;
+        if (__instance.IsLocalPlayerAboutToGetKilledByEnemy(other))
+            if (Helper.LocalPlayer != null) {
+                // instead let's make it chase the player, but never kill it.
+                if (__instance.currentBehaviourStateIndex is 0 or 1) {
+                    _ = __instance.Reflect().InvokeInternalMethod("ChaseLocalPlayer");
+                }
+                else {
+                    if (__instance.currentBehaviourStateIndex != 2 ||
+                        __instance.Reflect().GetInternalField<bool>("inLunge"))
+                        return false;
+                    __instance.transform.LookAt(Helper.LocalPlayer.transform.position);
+                    __instance.transform.localEulerAngles = new Vector3(0.0f, Helper.LocalPlayer.transform.eulerAngles.y, 0.0f);
+                    _ = __instance.Reflect().SetInternalField("inLunge", true);
+                    _ = __instance.Reflect().InvokeInternalMethod("EnterLunge");
+                }
+
+                return false;
+            }
+
+        return true;
+    }
```
I don't understand why you are not patching `KillPlayerServerRpc` instead. Did you test that it was this method that's causing the problem?

xAstroBoy replied:
yeah, each enemy fatal kill needs to be blocked to avoid getting the player bugged somewhere.

winstxnhdw replied:
It's bugged because of this line.

```cs
playerControllerB.inAnimationWithEnemy = this;
```

```cs
    public override void OnCollideWithPlayer(Collider other)
    {
        base.OnCollideWithPlayer(other);
        PlayerControllerB playerControllerB = MeetsStandardPlayerCollisionConditions(other, inKillAnimation);
        if (!(playerControllerB != null))
        {
            return;
        }

        Vector3 vector = Vector3.Normalize((base.transform.position + Vector3.up - playerControllerB.gameplayCamera.transform.position) * 100f);
        if (Physics.Linecast(base.transform.position + Vector3.up + vector * 0.5f, playerControllerB.gameplayCamera.transform.position, out var hitInfo, StartOfRound.Instance.collidersAndRoomMask, QueryTriggerInteraction.Ignore))
        {
            if (!(hitInfo.collider == debugCollider))
            {
                Debug.Log("Eyeless dog collide, linecast obstructed: " + hitInfo.collider.gameObject.name);
                debugCollider = hitInfo.collider;
            }
        }
        else if (currentBehaviourStateIndex == 3)
        {
            playerControllerB.inAnimationWithEnemy = this;
            KillPlayerServerRpc((int)playerControllerB.playerClientId);
        }
        else if (currentBehaviourStateIndex == 0 || currentBehaviourStateIndex == 1)
        {
            ChaseLocalPlayer();
        }
        else if (currentBehaviourStateIndex == 2 && !inLunge)
        {
            base.transform.LookAt(other.transform.position);
            base.transform.localEulerAngles = new Vector3(0f, base.transform.eulerAngles.y, 0f);
            inLunge = true;
            EnterLunge();
        }
    }
```

All you need to do is add a prefix that sets `currentBehaviourStateIndex` to not 3.

xAstroBoy replied:
Yeah, but id rather want the enemy do the lunge animation, which is why behaviour 3 is set.

winstxnhdw replied:
Lunge animation is 2 not 3.

xAstroBoy replied:
enum MouthDog {
    ROAMING = 0,
    SUSPICIOUS = 1,
    CHASE = 2,
    LUNGE = 3
}
no, is 3.

winstxnhdw replied:
The DLL literally says it's 2. 3 is only set AFTER entering a lunge.

xAstroBoy replied:
setting the RPC to 3 it goes in lunge animation.

winstxnhdw replied:
No, it doesn't. It just enrages the dog.

winstxnhdw replied:
```cs
            case 3:
                if (base.IsOwner)
                {
                    agent.speed -= Time.deltaTime * 5f;
                    if (!endingLunge && agent.speed < 1.5f && !inKillAnimation)
                    {
                        endingLunge = true;
                        lungeCooldown = 0.25f;
                        EndLungeServerRpc();
                    }
                }

                break;
```

It represents the end of a lunge.

winstxnhdw replied:
```cs
            case 2:
                if (!hasEnteredChaseModeFully)
                {
                    if (!startedChaseModeCoroutine)
                    {
                        startedChaseModeCoroutine = true;
                        StartCoroutine(enterChaseMode());
                    }

                    break;
                }

                ((RigConstraint<DampedTransformJob, DampedTransformData, DampedTransformJobBinder<DampedTransformData>>)(object)neckDampedTransform).weight = Mathf.Lerp(((RigConstraint<DampedTransformJob, DampedTransformData, DampedTransformJobBinder<DampedTransformData>>)(object)neckDampedTransform).weight, 0.2f, 8f * Time.deltaTime);
                if (!base.IsOwner)
                {
                    break;
                }

                if (base.IsOwner && roamPlanet.inProgress)
                {
                    StopSearch(roamPlanet);
                }

                if (!inLunge)
                {
                    lungeCooldown -= Time.deltaTime;
                    if (Vector3.Distance(base.transform.position, noisePositionGuess) < 4f && lungeCooldown <= 0f)
                    {
                        inLunge = true;
                        EnterLunge();
                        break;
                    }
                }

                agent.speed = Mathf.Clamp(agent.speed + Time.deltaTime, 13f, 18f);
                if (stunNormalizedTimer > 0f)
                {
                    agent.speed = 0f;
                }

                AITimer -= Time.deltaTime;
                if (AITimer <= 0f)
                {
                    AITimer = 3f;
                    suspicionLevel--;
                    if (Vector3.Distance(base.transform.position, agent.destination) < 3f)
                    {
                        SearchForPreviouslyHeardSound();
                    }

                    if (suspicionLevel <= 8)
                    {
                        SwitchToBehaviourState(1);
                    }
                }

                break;
```

2 is the lunge.

xAstroBoy replied:
2 makes the dog run, 3 makes it do the ambush animation, and if that hits a player is calling the kill RPC, you can literally test it out with possessionmod!

winstxnhdw replied:
I did test with PossessionMod. 3 does makes it do a roar. Not a lunge.

xAstroBoy replied:
Feel free to make any changes you think are the best ones.

winstxnhdw replied:
It's alright. I can wait.

winstxnhdw commented on lc-hax/Scripts/Modules/Possession/Controllers/NutcrackerController.cs:
```diff
@@ -6,7 +6,13 @@ enum NutCrackerState {
+    private bool isSecondarySkillActive = false;
+
+    public void OnMovement(NutcrackerEnemyAI enemyInstance, bool isMoving, bool isSprinting) {
+        if (isMoving || isSprinting) {
+            enemyInstance.SetBehaviourState(NutCrackerState.WALKING);
+        }
```
```suggestion
        if (!isMoving && !isSprinting) return;
        enemyInstance.SetBehaviourState(NutCrackerState.WALKING);
```

winstxnhdw commented on lc-hax/Scripts/Helpers/Enemies.cs:
```diff
@@ -19,6 +20,61 @@ internal static partial class Helper {
+        return playerControllerB != null && playerControllerB.IsSelf();
+    }
+
+    private static Dictionary<string, GameObject>? _AllSpawnableEnemies;
+
+    internal static Dictionary<string, GameObject> AllSpawnableEnemies {
+        get {
+            if (_AllSpawnableEnemies == null) {
+                _AllSpawnableEnemies = [];
+
+                HashSet<(string Name, GameObject Prefab)> uniqueEnemies = [];
+
+                SelectableLevel[] levels = Resources.FindObjectsOfTypeAll<SelectableLevel>();
+
+                foreach (SelectableLevel level in levels) {
+                    List<List<SpawnableEnemyWithRarity>> enemyCollections = [
+                        level.Enemies,
+                        level.OutsideEnemies,
+                        level.DaytimeEnemies
+                    ];
+
+                    foreach (List<SpawnableEnemyWithRarity> collection in enemyCollections) {
+                        foreach (SpawnableEnemyWithRarity enemy in collection) {
+                            if (enemy.enemyType.enemyName.Contains("Docile Locust Bees", StringComparison.OrdinalIgnoreCase)) continue;
+                            if (enemy.enemyType.enemyName.Contains("Manticoil", StringComparison.OrdinalIgnoreCase)) continue;
+                            _ = uniqueEnemies.Add((enemy.enemyType.enemyName, enemy.enemyType.enemyPrefab));
+                        }
+                    }
+                }
+
+                foreach ((string Name, GameObject Prefab) in uniqueEnemies) {
+                    _ = _AllSpawnableEnemies.TryAdd(Name, Prefab);
+                }
+            }
+
+            return _AllSpawnableEnemies;
+        }
+    }
```
What's the point of this? Where will we be reusing this?

xAstroBoy replied:
is being used in /spawn and /enemies.
it does it once as it parses all prefabs and names from all loaded levels , including modded ones and modded enemies.

winstxnhdw replied:
> that was the last one, im enchancing /spawn to cover all enemies of all levels, not just the ones based off that level, 

Didn't you mention this? If you could spawn all enemies across levels, why do we need to check for this?

xAstroBoy replied:
Oh, that's because each SelectableLevel has the prefabs registered there, i dont know any other way to parse it.

winstxnhdw replied:
```cs
Resources.FindObjectsOfTypeAll<EnemyType>().Select(enemyType => enemyType.enemyPrefab);
```

xAstroBoy replied:
Fixed

winstxnhdw replied:
I give up.

xAstroBoy replied:
> I give up.

you really want to rely always on Resources.FindObjectsOfTypeAll and not on a filled dictionary that will fill once?

winstxnhdw replied:
Nope. I am more amazed that you would rather use 2 for-loops instead of using `Where` and `Select`.

xAstroBoy replied:
> Nope. I am more amazed that you would rather use 2 for-loops instead of using `Where` and `Select`.

??? is just one foreach!

winstxnhdw replied:
You're right, but you could just return directly if you did it with method chaining rather than this mess.

```cs

            if (_AllSpawnableEnemies == null) {
                _AllSpawnableEnemies = [];
                foreach (EnemyType enemy in Resources.FindObjectsOfTypeAll<EnemyType>()) {
                    if (enemy.enemyName.Contains("Docile Locust Bees", StringComparison.OrdinalIgnoreCase)) continue;
                    if (enemy.enemyName.Contains("Manticoil", StringComparison.OrdinalIgnoreCase)) continue;
                    _ = _AllSpawnableEnemies.TryAdd(enemy.enemyName, enemy.enemyPrefab);
                }
            }

            return _AllSpawnableEnemies;

```

winstxnhdw commented on lc-hax/Scripts/Patches/UnlockableSuitPatch.cs:
You don't even need a patch for this. The fields are all public.

winstxnhdw commented:
I am not changing `sell`. It works better as it was.

xAstroBoy commented:
> I am not changing `sell`. It works better as it was.

issue is your algoritm does not work, i adapted to work with your grab logic and it did do nothing, unless i sell all stuff.

winstxnhdw commented:
No idea what do you mean by not working. I played the entire night and used the sell command many times. Worked perfectly.

xAstroBoy commented:
> No idea what do you mean by not working. I played the entire night and used the sell command many times. Worked perfectly.

then just take the modified PlaceItemOnCounter and adapt it to work with .GrabObject as i did, that way when you type /sell it will not show floating items randomly flying to the desk, it will show you placing them there.

winstxnhdw commented:
I want it to fly to the desk. Why would I want others to know it was me who sold the items?

xAstroBoy commented:
> I want it to fly to the desk. Why would I want others to know it was me who sold the items?

oh, then just adapt PlaceItemsOnDesk , it works either way without a player holding it as you will force place the item there.

winstxnhdw commented:
The `main` branch version of `sell` is atomic. But you want to switch to `GrabObject` which requires async code, is less performant, and harder to maintain. There is no good reason to use it over what we already have.

xAstroBoy commented:
> The `main` branch version of `sell` is atomic. But you want to switch to `GrabObject` which requires async code, is less performant, and harder to maintain. There is no good reason to use it over what we already have.

i mean did you try just invoking the AddObjectToDeskServerRPC without even holding the item only?

winstxnhdw commented:
Not sure what you mean.

xAstroBoy commented:
> Not sure what you mean.

nevermind, i discarded all changes about the sell command, invoking the RPC only is not enough.

winstxnhdw commented:
Yes, that’s how grabbables work in this game. You need to first ‘grab’ them to gain ownership of the item.

winstxnhdw commented on lc-hax/Scripts/Helpers/Color.cs:
```diff
@@ -0,0 +1,170 @@
+        internal static Color Aquamarine { get; } = new(0.4980392f, 1f, 0.8313726f, 1f);
+        internal static Color Aqua { get; } = new(0f, 1f, 1f, 1f);
+        internal static Color AntiqueWhite { get; } = new(0.9803922f, 0.9215686f, 0.8431373f, 1f);
+        internal static Color AliceBlue { get; } = new(0.9411765f, 0.972549f, 1f, 1f);
+        internal static Color Transparent { get; } = new(1f, 1f, 1f, 1f);
+        internal static Color BurlyWood { get; } = new(0.8705882f, 0.7215686f, 0.5294118f, 1f);
+        internal static Color DarkOliveGreen { get; } = new(0.3333333f, 0.4196078f, 0.1843137f, 1f);
+        internal static Color CornflowerBlue { get; } = new(0.3921569f, 0.5843138f, 0.9294118f, 1f);
+        internal static Color DarkOrchid { get; } = new(0.6f, 0.1960784f, 0.8f, 1f);
+        internal static Color Khaki { get; } = new(0.9411765f, 0.9019608f, 0.5490196f, 1f);
+        internal static Color Ivory { get; } = new(1f, 1f, 0.9411765f, 1f);
+        internal static Color DarkOrange { get; } = new(1f, 0.5490196f, 0f, 1f);
+        internal static Color Indigo { get; } = new(0.2941177f, 0f, 0.509804f, 1f);
+        internal static Color IndianRed { get; } = new(0.8039216f, 0.3607843f, 0.3607843f, 1f);
+        internal static Color HotPink { get; } = new(1f, 0.4117647f, 0.7058824f, 1f);
+        internal static Color Honeydew { get; } = new(0.9411765f, 1f, 0.9411765f, 1f);
+        internal static Color GreenYellow { get; } = new(0.6784314f, 1f, 0.1843137f, 1f);
+        internal static Color Green { get; } = new(0f, 0.5019608f, 0f, 1f);
+        internal static Color Gray { get; } = new(0.5019608f, 0.5019608f, 0.5019608f, 1f);
+        internal static Color Goldenrod { get; } = new(0.854902f, 0.6470588f, 0.1254902f, 1f);
+        internal static Color GhostWhite { get; } = new(0.972549f, 0.972549f, 1f, 1f);
+        internal static Color Gainsboro { get; } = new(0.8627451f, 0.8627451f, 0.8627451f, 1f);
+        internal static Color Fuchsia { get; } = new(1f, 0f, 1f, 1f);
+        internal static Color Gold { get; } = new(1f, 0.8431373f, 0f, 1f);
+        internal static Color FloralWhite { get; } = new(1f, 0.9803922f, 0.9411765f, 1f);
+        internal static Color DarkRed { get; } = new(0.5450981f, 0f, 0f, 1f);
+        internal static Color DarkSalmon { get; } = new(0.9137255f, 0.5882353f, 0.4784314f, 1f);
+        internal static Color DarkSeaGreen { get; } = new(0.5607843f, 0.7372549f, 0.5450981f, 1f);
+        internal static Color ForestGreen { get; } = new(0.1333333f, 0.5450981f, 0.1333333f, 1f);
+        internal static Color DarkSlateGray { get; } = new(0.1843137f, 0.3098039f, 0.3098039f, 1f);
+        internal static Color DarkTurquoise { get; } = new(0f, 0.8078431f, 0.8196079f, 1f);
+        internal static Color DarkSlateBlue { get; } = new(0.282353f, 0.2392157f, 0.5450981f, 1f);
+        internal static Color DeepPink { get; } = new(1f, 0.07843138f, 0.5764706f, 1f);
+        internal static Color DeepSkyBlue { get; } = new(0f, 0.7490196f, 1f, 1f);
+        internal static Color DimGray { get; } = new(0.4117647f, 0.4117647f, 0.4117647f, 1f);
+        internal static Color DodgerBlue { get; } = new(0.1176471f, 0.5647059f, 1f, 1f);
+        internal static Color Firebrick { get; } = new(0.6980392f, 0.1333333f, 0.1333333f, 1f);
+        internal static Color DarkViolet { get; } = new(0.5803922f, 0f, 0.827451f, 1f);
+    }
+}
```
Where did you steal this from?

xAstroBoy replied:
That's from AstroClient, my old VRChat client i did awhile ago.

winstxnhdw replied:
You handwrote all of these colours?

xAstroBoy replied:
more like i made a utility to convert the entire System.Colors to unity and then i converted the results into a struct

xAstroBoy replied:
ESP identification looks way better than the unity stock color options.
i also include a way to convert Hex to Unity color directly, but is better if you do it  into a property or field and go from there.

xAstroBoy replied:
<img width="1140" alt="image" src="https://github.com/winstxnhdw/lc-hax/assets/92211398/a5bc4fcf-55f3-42fb-b4db-e7a2d859aac3">

xAstroBoy commented:
@winstxnhdw i hope you agree the ESP looks better.
i fixed the ESP rendering unexistent players as well made it change color if a player died and a body exists.

xAstroBoy commented:
@winstxnhdw now the body shows which player is from

winstxnhdw commented:
Seems like your PlaySFX is not server-side.

xAstroBoy commented:
> Seems like your PlayDropSFX is server-side only.

what do you mean?

winstxnhdw commented:
I think I meant to type ‘client-side’ only.

xAstroBoy commented:
> I think I meant to type ‘client-side’ only.

how so? im doing it exactly how the game plays it when you drop, i couldn't find a rpc for it

winstxnhdw commented:
My friends can’t hear it. I’ll be removing it shortly.

xAstroBoy commented:
> My friends can’t hear it. I’ll be removing it shortly.

but they can hear the rest of the specific interactions?

winstxnhdw commented:
Haven’t tested those, so I’ll leave it until I test them.
